crates/bevy_ecs/src/component.rs

```diff
@@ -347,21 +348,29 @@ pub struct ComponentTicks {
 impl ComponentTicks {
     #[inline]
     pub fn is_added(&self, last_change_tick: u32, change_tick: u32) -> bool {
-        // The comparison is relative to `change_tick` so that we can detect changes over the whole
-        // `u32` range. Comparing directly the ticks would limit to half that due to overflow
-        // handling.
-        let component_delta = change_tick.wrapping_sub(self.added);
-        let system_delta = change_tick.wrapping_sub(last_change_tick);
+        // The comparison is relative to the world tick (`change_tick`) so that we can use
+        // the full `u32` range. Comparing the system and component ticks directly would limit the
+        // maximum delta to `u32::MAX / 2` because of wraparound.
+        let ticks_since_insert = change_tick
+            .wrapping_sub(self.added)
+            .min(MAX_TICK_DELTA);
+        let ticks_since_system = change_tick
+            .wrapping_sub(last_change_tick)
+            .min(MAX_TICK_DELTA);
```

Comment:
I don´t think that taking the minimums is the best way to go. The is_added and is_changed functions need to be as optimized as possible to reduce the cost of change detection, and this change targets a very uncommon case.
Moreover, I thought that change older than MAX_TICK_DELTA were already ignored in practice, but I´ll need to check.

Reply:
In practice this would be an extreme edge case, but I don't think changes older than `MAX_TICK_DELTA` are ignored. We allow things to be in the `MAX_TICK_DELTA..=u32::MAX` range between scans.

If you have a system that last ran `MAX_TICK_DELTA - 1` ticks ago and a component that last changed `MAX_TICK_DELTA - 2` ticks ago, a scan happening at that moment would not modify them but they'll continue to get older before the next scan. If that system runs again before the next scan, the change will be detected because `ticks_since_system > ticks_since_change`, even though both are older than `MAX_TICK_DELTA`. 

That's unreliable, though. The first scan happening two ticks later would reverse the outcome.

Reply:
That said, I think we can move the clamp out of `is_added` and `is_changed`.

If we clamp a system's `last_change_tick` value every time it runs, `ticks_since_system > ticks_since_change` should return `false` even if changes are older than `MAX_TICK_DELTA`.

Reply:
> If we clamp a system's `last_change_tick` value every time it runs, `ticks_since_system > ticks_since_change` should return false even if changes are older than `MAX_TICK_DELTA`.

So I really don't like this after all. I'd have to modify the `get_param` of every system param (except `SystemChangeTick`).

Taking the minimum only adds 5 instructions when compiled in release mode (https://rust.godbolt.org/z/G9crazb5d) and I see identical benchmark results between bevy main and this PR.

I strongly prefer leaving this as is.

![image](https://user-images.githubusercontent.com/51241057/154753067-bfdfc175-cc74-411f-8f7b-5ff6e9543869.png)

Reply:
Marking as resolved. If anyone sees something I missed, feel free to re-open this discussion.

Reply:
Reading it again, I don't actually understand how it works in the common case where both `change_tick.wrapping_sub(self.added)` and `change_tick.wrapping_sub(last_change_tick)` are lower than `MAX_CHANGE_AGE`. It seems they would be both set to `MAX_CHANGE_AGE`, thus equal, and return false regardless of which is older?

Reply:
If both ages are lower than `MAX_CHANGE_AGE`, the `min` comparison will return their actual ages.

When something becomes older than `MAX_CHANGE_AGE`, we can't make any guarantees about how far it'll get before the next `check_tick` scan happens (we don't know when it will happen), so we clamp to be deterministic.

Reply:
Right sorry, I was confused 😅 Since there is no significant change in the benchmark, I'm ok with it.